custom_components/spotcast/helpers.py

```diff
@@ -103,117 +107,324 @@ async def run(*args, loop=None, executor=None, **kwargs):
+    for item in [playlistName, showName, episodeName, audiobookName]:
+        if not is_empty_str(item):
+            search.append(item)
+
+    return " ".join(search)
+
+
+# "album", "artist", "playlist", "track", "show", "episode", "audiobook"
+def get_types_string(
+    artistName: str,
+    albumName: str,
+    trackName: str,
+    playlistName: str,
+    showName: str,
+    episodeName: str,
+    audiobookName: str,
+) -> str:
+    types = []
+    if not is_empty_str(artistName):
+        types.append("artist")
+    if not is_empty_str(albumName):
+        types.append("album")
+    if not is_empty_str(trackName):
+        types.append("track")
+    if not is_empty_str(playlistName):
+        types.append("playlist")
+    if not is_empty_str(showName):
+        types.append("show")
+    if not is_empty_str(episodeName):
+        types.append("episode")
+    if not is_empty_str(audiobookName):
+        types.append("audiobook")
+    return ",".join(types)
+
+
+def get_search_results(
+    spotify_client: spotipy.Spotify,
+    limit: int = 10,
+    country: str = None,
+    artistName: str = None,
+    albumName: str = None,
+    playlistName: str = None,
+    trackName: str = None,
+    showName: str = None,
+    episodeName: str = None,
+    audiobookName: str = None,
+    genreName: str = None,
+):
+    _LOGGER.debug("using search query to find uri")
+    searchResults = []
+
+    if (
+        not is_empty_str(artistName)
+        and len(
+            list(
+                filter(
+                    lambda x: not is_empty_str(x),
+                    [
+                        albumName,
+                        playlistName,
+                        trackName,
+                        showName,
+                        episodeName,
+                        audiobookName,
+                        genreName,
+                    ],
+                )
             )
+        )
+        == 0
+    ):
+        searchResults = get_top_tracks(artistName, spotify_client)
+        _LOGGER.debug("Playing top tracks for artist: %s", searchResults[0]["name"])
+        return searchResults
+    else:
+        searchString = get_search_string(
+            artistName=artistName,
+            albumName=albumName,
+            trackName=trackName,
+            genreName=genreName,
+            playlistName=playlistName,
+            showName=showName,
+            episodeName=episodeName,
+            audiobookName=audiobookName,
+        )
+
+        searchTypes = get_types_string(
+            artistName=artistName,
+            albumName=albumName,
+            trackName=trackName,
+            playlistName=playlistName,
+            showName=showName,
+            episodeName=episodeName,
+            audiobookName=audiobookName,
+        )
+        searchResults = spotify_client.search(
+            q=searchString, limit=limit, offset=0, type=searchTypes, market=country
+        )
+
+        compiledResults = []
```

Comment:
There is probably a more elegant way to write this part, I was in a rush at this point to get something working. Basically, the spotify api returns different entities nested in different locations. The original code would return only "tracks", but we want to glom together all types of entities.